web3/middleware/signing.py

```diff
@@ -0,0 +1,64 @@
+from eth_keys.datatypes import PrivateKey
+
+from web3.utils.datastructures import HexBytes
+# from web3.utils.signing import LocalAccount
+
+import logging
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def construct_transaction_signing_middleware(private_key):
+    """Intercept a transaction in flight and sign it.
+
+    Args:
+        private_key (TYPE): Private key used to sign the transaction.
+                            Accepts the following formats:
+                            - `PrivateKey` object from `eth-keys`
+                            - `LocalAccount` object from `web3.eth.account`
+                            - raw bytes
+                            - TODO: hex encoded string
+
+
+    Returns:
+        TYPE: Description
+    """
+
+    # Check if the private_key is either
+    # - eth_keys.keys.PrivateKey
+    # - web3.eth.account.create().privateKey -- can just check that its HexBytes
+    # - bytes
+    # TODO: - hexbytes
+    valid_key_types = (PrivateKey, HexBytes, bytes)
+    _private_key = None
+
+    for t in valid_key_types:
+        if isinstance(private_key, t):
+            # Might also want to check that bytes length == 32
+            _private_key = private_key
+            break
+
+    def transaction_signing_middleware(make_request, web3):
+        def middleware(method, params):
+            # Only operate on the `eth.sendTransaction` method
+            # Only operate on if the private key matches the public key in the transaction
+            # Note: params == transaction in this case.
+            if method is not 'eth_sendTransaction':
+                return make_request(method, params)
+            if params.get('from') != web3.eth.account.privateKeyToAccount(_private_key):
```

Comment:
since `privateKeyToAccount` involves some ECDSA operations, we should probably pre-compute the address and do a comparison against the pre-computed address.

Also, while we expect these two values to both be checksummed, it is probably a good idea to use `eth_utils.is_same_address` here since it gracefully handles comparisons of addresses which are *equal* but who's strings are not equal.

Reply:
fixed 👍 .  `eth_utils` is definitely the way to go.